### ISSUE TITLE
[Bug] Cannot embed IG post via CMS

### ISSUE THREAD

**Issue by prestonlimlianjie** [isomerpages/isomercms-frontend#893]
### Problem

Users are unable to embed Instagram posts by copying the content from Instagram's embed instructions and pasting it into the CMS.

![Screenshot 2022-05-24 at 5 39 57 PM](https://user-images.githubusercontent.com/19917616/170001716-2506f075-7c77-4433-9726-c117c3cabfd9.png)

### Details
Instagram provides an embed snippet as follows:

```
<blockquote class="instagram-media" data-instgrm-permalink="https://www.instagram.com/p/Cduu7D9JX_M/?utm_source=ig_embed&amp;utm_campaign=loading" data-instgrm-version="14" style=" background:#FFF; border:0; border-radius:3px; box-shadow:0 0 1px 0 rgba(0,0,0,0.5),0 1px 10px 0 rgba(0,0,0,0.15); margin: 1px; max-width:540px; min-width:326px; padding:0; width:99.375%; width:-webkit-calc(100% - 2px); width:calc(100% - 2px);"><div style="padding:16px;"> <a href="https://www.instagram.com/p/Cduu7D9JX_M/?utm_source=ig_embed&amp;utm_campaign=loading" style=" background:#FFFFFF; line-height:0; padding:0 0; text-align:center; text-decoration:none; width:100%;" target="_blank"> <div style=" display: flex; flex-direction: row; align-items: center;"> <div style="background-color: #F4F4F4; border-radius: 50%; flex-grow: 0; height: 40px; margin-right: 14px; width: 40px;"></div> <div style="display: flex; flex-direction: column; flex-grow: 1; justify-content: center;"> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; margin-bottom: 6px; width: 100px;"></div> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; width: 60px;"></div></div></div><div style="padding: 19% 0;"></div> <div style="display:block; height:50px; margin:0 auto 12px; width:50px;"><svg width="50px" height="50px" viewBox="0 0 60 60" version="1.1" xmlns="https://www.w3.org/2000/svg" xmlns:xlink="https://www.w3.org/1999/xlink"><g stroke="none" stroke-width="1" fill="none" fill-rule="evenodd"><g transform="translate(-511.000000, -20.000000)" fill="#000000"><g><path d="M556.869,30.41 C554.814,30.41 553.148,32.076 553.148,34.131 C553.148,36.186 554.814,37.852 556.869,37.852 C558.924,37.852 560.59,36.186 560.59,34.131 C560.59,32.076 558.924,30.41 556.869,30.41 M541,60.657 C535.114,60.657 530.342,55.887 530.342,50 C530.342,44.114 535.114,39.342 541,39.342 C546.887,39.342 551.658,44.114 551.658,50 C551.658,55.887 546.887,60.657 541,60.657 M541,33.886 C532.1,33.886 524.886,41.1 524.886,50 C524.886,58.899 532.1,66.113 541,66.113 C549.9,66.113 557.115,58.899 557.115,50 C557.115,41.1 549.9,33.886 541,33.886 M565.378,62.101 C565.244,65.022 564.756,66.606 564.346,67.663 C563.803,69.06 563.154,70.057 562.106,71.106 C561.058,72.155 560.06,72.803 558.662,73.347 C557.607,73.757 556.021,74.244 553.102,74.378 C549.944,74.521 548.997,74.552 541,74.552 C533.003,74.552 532.056,74.521 528.898,74.378 C525.979,74.244 524.393,73.757 523.338,73.347 C521.94,72.803 520.942,72.155 519.894,71.106 C518.846,70.057 518.197,69.06 517.654,67.663 C517.244,66.606 516.755,65.022 516.623,62.101 C516.479,58.943 516.448,57.996 516.448,50 C516.448,42.003 516.479,41.056 516.623,37.899 C516.755,34.978 517.244,33.391 517.654,32.338 C518.197,30.938 518.846,29.942 519.894,28.894 C520.942,27.846 521.94,27.196 523.338,26.654 C524.393,26.244 525.979,25.756 528.898,25.623 C532.057,25.479 533.004,25.448 541,25.448 C548.997,25.448 549.943,25.479 553.102,25.623 C556.021,25.756 557.607,26.244 558.662,26.654 C560.06,27.196 561.058,27.846 562.106,28.894 C563.154,29.942 563.803,30.938 564.346,32.338 C564.756,33.391 565.244,34.978 565.378,37.899 C565.522,41.056 565.552,42.003 565.552,50 C565.552,57.996 565.522,58.943 565.378,62.101 M570.82,37.631 C570.674,34.438 570.167,32.258 569.425,30.349 C568.659,28.377 567.633,26.702 565.965,25.035 C564.297,23.368 562.623,22.342 560.652,21.575 C558.743,20.834 556.562,20.326 553.369,20.18 C550.169,20.033 549.148,20 541,20 C532.853,20 531.831,20.033 528.631,20.18 C525.438,20.326 523.257,20.834 521.349,21.575 C519.376,22.342 517.703,23.368 516.035,25.035 C514.368,26.702 513.342,28.377 512.574,30.349 C511.834,32.258 511.326,34.438 511.181,37.631 C511.035,40.831 511,41.851 511,50 C511,58.147 511.035,59.17 511.181,62.369 C511.326,65.562 511.834,67.743 512.574,69.651 C513.342,71.625 514.368,73.296 516.035,74.965 C517.703,76.634 519.376,77.658 521.349,78.425 C523.257,79.167 525.438,79.673 528.631,79.82 C531.831,79.965 532.853,80.001 541,80.001 C549.148,80.001 550.169,79.965 553.369,79.82 C556.562,79.673 558.743,79.167 560.652,78.425 C562.623,77.658 564.297,76.634 565.965,74.965 C567.633,73.296 568.659,71.625 569.425,69.651 C570.167,67.743 570.674,65.562 570.82,62.369 C570.966,59.17 571,58.147 571,50 C571,41.851 570.966,40.831 570.82,37.631"></path></g></g></g></svg></div><div style="padding-top: 8px;"> <div style=" color:#3897f0; font-family:Arial,sans-serif; font-size:14px; font-style:normal; font-weight:550; line-height:18px;">View this post on Instagram</div></div><div style="padding: 12.5% 0;"></div> <div style="display: flex; flex-direction: row; margin-bottom: 14px; align-items: center;"><div> <div style="background-color: #F4F4F4; border-radius: 50%; height: 12.5px; width: 12.5px; transform: translateX(0px) translateY(7px);"></div> <div style="background-color: #F4F4F4; height: 12.5px; transform: rotate(-45deg) translateX(3px) translateY(1px); width: 12.5px; flex-grow: 0; margin-right: 14px; margin-left: 2px;"></div> <div style="background-color: #F4F4F4; border-radius: 50%; height: 12.5px; width: 12.5px; transform: translateX(9px) translateY(-18px);"></div></div><div style="margin-left: 8px;"> <div style=" background-color: #F4F4F4; border-radius: 50%; flex-grow: 0; height: 20px; width: 20px;"></div> <div style=" width: 0; height: 0; border-top: 2px solid transparent; border-left: 6px solid #f4f4f4; border-bottom: 2px solid transparent; transform: translateX(16px) translateY(-4px) rotate(30deg)"></div></div><div style="margin-left: auto;"> <div style=" width: 0px; border-top: 8px solid #F4F4F4; border-right: 8px solid transparent; transform: translateY(16px);"></div> <div style=" background-color: #F4F4F4; flex-grow: 0; height: 12px; width: 16px; transform: translateY(-4px);"></div> <div style=" width: 0; height: 0; border-top: 8px solid #F4F4F4; border-left: 8px solid transparent; transform: translateY(-4px) translateX(8px);"></div></div></div> <div style="display: flex; flex-direction: column; flex-grow: 1; justify-content: center; margin-bottom: 24px;"> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; margin-bottom: 6px; width: 224px;"></div> <div style=" background-color: #F4F4F4; border-radius: 4px; flex-grow: 0; height: 14px; width: 144px;"></div></div></a><p style=" color:#c9c8cd; font-family:Arial,sans-serif; font-size:14px; line-height:17px; margin-bottom:0; margin-top:8px; overflow:hidden; padding:8px 0 7px; text-align:center; text-overflow:ellipsis; white-space:nowrap;"><a href="https://www.instagram.com/p/Cduu7D9JX_M/?utm_source=ig_embed&amp;utm_campaign=loading" style=" color:#c9c8cd; font-family:Arial,sans-serif; font-size:14px; font-style:normal; font-weight:normal; line-height:17px; text-decoration:none;" target="_blank">A post shared by SPCA Singapore Adoption (@spcasgadopt)</a></p></div></blockquote><script async src="//www.instagram.com/embed.js"></script>
```

It's quite a long snippet, but the key issue is that there is a `<script async src>` call to `www.instagram.com`.
```
<script async src="//www.instagram.com/embed.js"></script>
```

This is prevented by the `checkCSP()` function.

### Solution

We could load the Instagram embed.js file in each EditPage view by default, and get users to modify the embed instructions before pasting it into the editor.

**Comment by dcshzj** [comment]
### Additional considerations

1. When the user is creating/editing the page and adds the Instagram embedded post, the embed.js file will need to be loaded in the EditPage view, so that the embedded post can be loaded in the preview pane.
2. When the user saves the page with the Instagram embedded post, the page on the deployed Isomer site should also load the embed.js file so that the embedded post can be loaded as expected.
3. If there are no embedded Instagram posts on the page, the embed.js file should not be loaded.

### Proposed solution 1: Introduce a whitelist for &lt;script&gt; tags

Currently, Netlify sends a `content-security-policy` header which specifies the list of allowed domains. As of 2022-08-11, the domains that are currently enabled are:

```
https://assets.dcube.cloud
https://*.wogaa.sg
https://assets.adobedtm.com
https://www.google-analytics.com
https://cdnjs.cloudflare.com
https://va.ecitizen.gov.sg
https://*.cloudfront.net
https://printjs-4de6.kxcdn.com
https://unpkg.com
https://wogadobeanalytics.sc.omtrdc.net
https://connect.facebook.net
https://graph.facebook.com
https://facebook.com
https://www.facebook.com
https://www.googletagmanager.com
https://*.licdn.com
https://webchat.vica.gov.sg
https://vica.gov.sg
https://www.google.com/recaptcha/
https://www.gstatic.com/recaptcha/
https://static.zdassets.com
https://ekr.zdassets.com
https://*.zendesk.com
https://*.zopim.com
https://www.instagram.com
wss://*.zendesk.com
wss://*.zopim.com
```

We can selectively enable the &lt;script&gt; tags that matches this list of domains (excluding the `wss://` ones). This implementation satisfies points 2 and 3 in the "Additional considerations" section above, as the scripts are only loaded when needed.

However, some additional investigations are needed:

1. The input undergoes further sanitisation before it can be saved using DOMPurify. Is it possible to override this sanitisation to allow the &lt;script&gt; tags with the source matching the above list of domains? Should we allow this, or should we treat all user input as dangerous?
2. Does the rendered preview allow for &lt;script&gt; tags to load during preview?

### Proposed solution 2: Introduce a special "insert media" function in the Markdown editor

We can extend the existing Markdown editor to support inserting embedded Instagram posts [through extending the toolbar](https://github.com/sparksuite/simplemde-markdown-editor#toolbar-icons). This can potentially allow us to improve the user experience by displaying a modal for the user to search/specify the Instagram post to display, and potentially extend this functionality to embed other media such as YouTube videos.

However, this requires us to extend the Markdown syntax to introduce custom tags, which we can use to specify to the backend to perform some pre-processing to add the &lt;script&gt; tag to load the embed.js file, before saving the changes into the GitHub repository.

The following would be a potential user flow:

1. User selects the "Insert media" function on the toolbar.
2. A modal appears for the user to specify the Instagram post to embed (can be further extended to other forms of media in the future).
3. Once the user is satisfied with the display, the user clicks on "Insert". A special tag (e.g. `<Instagram />`) will be added to the Markdown editor to indicate the embedded Instagram post. The preview pane will also load the embedded post as expected.
4. When the user saves the page, the IsomerCMS backend will perform some processing of the page text, transforming the special tag into the actual code that should be added into the page. The processed text will then be saved into the GitHub repository.
5. Whenever the user edits the same page in the future, the actual code will be parsed and transformed back into the special tag syntax.

This proposed implementation satisfies points 2 and 3 in the "Additional considerations" section above, as the scripts are only loaded when needed. The same problem on whether the rendered preview allow for &lt;script&gt; tags to load during preview exists.

### Proposed solution 3: Introduce additional page configuration settings to selectively load embed.js

The IsomerCMS can potentially have an additional "Page settings" section that allows the user to configure page-specific settings. In this case, a toggle can be introduced to allow the user to specify if the page contains any embedded Instagram posts. If the toggle is enabled, then the embed.js file will be added into the page to be loaded.

The following would be a potential user flow:

1. The user copies the code to embed Instagram posts and pastes it into the Markdown editor.
2. The Instagram post does not load properly due to the missing embed.js script. The IsomerCMS frontend can potentially provide the user with a non-intrusive prompt to enable the Instagram toggle.
3. The user can enable the toggle to indicate that the page contains an embedded Instagram post.
4. The Instagram post now loads correctly as expected.
5. If the Instagram post is removed, the IsomerCMS frontend can either disable the Instagram toggle automatically, or provide the user with a non-intrusive prompt to disable the Instagram toggle.

Thus, when the user adds the code to embed Instagram posts, the &lt;script&gt; tag will be removed, but the Instagram post will still be loaded if the toggle is enabled. This proposed implementation can satisfy point 2 in the "Additional considerations" section above, but it may be possible that the user leaves the toggle enabled without having any Instagram posts embedded in the page. In such a case, it should be possible for the IsomerCMS frontend to automatically disable the toggle, if it detects that the page content does not contain any Instagram URLs.

This implementation can also open up the possibility of introducing more page-specific settings that the user can configure. However, some additional investigations are needed:

1. Does the rendered preview allow for &lt;script&gt; tags to load during preview?
2. Should this toggle be added to the user interface, or should embed.js be automatically added/removed transparently from the user?

**Comment by dcshzj** [comment]
Fixed in #1019, will be released together with the other identity phase 2 features.
